### PR TITLE
fix: request promise json

### PR DESCRIPTION
**Pull request by ludeknovy**
Fix for https://github.com/RuntimeTools/OpenAPIValidators/issues/80

### CODE REVIEW

**Comment by rwalle61** [comment]
Thanks for this! I'll review this properly as soon as I can.

My initial thoughts are:

I can't remember why we had the `replace` originally - I thought I put it in specifically for `request-promise`. We have [a test for this](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/differentRequestModules.test.js#L356) in our sister package `chai-openapi-response-validator`, which I'll copy into this repo when I get time.

Until then, if you copy this change to [the right place in `chai-openapi-response-validator`](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/RequestPromiseResponse.js#L7) then the tests will confirm it's safe to remove the `replace`.

(Sorry for the discrepancy and duplication between `jest-openapi` and `chai-openapi-response-validator`. I would love to sync the shared `openapi-validator` dependency, but haven't found a way to do so without breaking our `stryker` test infrastructure)

**Comment by rwalle61** [comment]
Haha I was about to comment asking for an update to the chai package. 

Thanks for this! I'll review this properly as soon as I can.

My initial thoughts are:

Thanks for the unit test. I think it'd be good also to add an integration test recreating this error. You can add one [in this test block](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/differentRequestModules.test.js#L356)

I thought I put the `replace` in for `request-promise`, but can't remember why exactly. The tests will confirm it's safe to remove `replace`


(Sorry for not syncing `jest-openapi` and `chai-openapi-response-validator` automatically - I haven't found a way to do so without breaking our `stryker` test infrastructure)

**Comment by ludeknovy** [comment]
I spent like 20min trying to figure out why `replace` was needed. And I came to decision it was not needed after all.
But I might be wrong :)

**Comment by rwalle61** [comment]
I introduced `replace` [here](https://github.com/RuntimeTools/OpenAPIValidators/pull/23/commits/23fa7056f71d5be51f19f223bad0a3d6d983ff3b#diff-d86a59ede7d999db4b7bc43cb25a1c11R19), but can't tell why exactly! 

At the end of the day the integration tests for `chai-openapi-response-validator` and `jest-openapi` capture all the functionality we support. So if the tests pass, I'm happy with our code.

That's why it'd be good to have an integration test recreating the error you described in https://github.com/RuntimeTools/OpenAPIValidators/issues/80. I'm happy for you to just add that test case for `chai-openapi-response-validator`, and I'll sync everything up this weekend.

Once that's in, I'm happy to merge

Thanks again!

**Comment by ludeknovy** [comment]
Ok, I will look at the integration tests during this weekend.

**Comment by rwalle61** [comment]
btw, I think I remember why `replace` is needed. I accidentally wrote [this `request-promise` test](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/differentRequestModules.test.js#L366) wrong  - it calls `supertest` but should actually be calling `requestPromise` (like the test below). 

When I corrected the test and ran it, I saw that the string `res.body` is `'"res.body is a string"'`. I think I used `replace` to handle this case.

Looks like we will a solution compatable with both this case and your new test case covering the bug you found

File: packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/RequestPromiseResponse.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ class RequestPromiseResponse extends AbstractResponse {
   constructor(res) {
     super(res);
     this.status = res.statusCode;
-    this.body = res.body.replace(/"/g, '');
+    this.body = res.request.headers.accept === 'application/json' ? res.body : res.body.replace(/"/g, '');
```
Comment (rwalle61) [inline packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/RequestPromiseResponse.js:7]:
```suggestion
    this.body = res.request._json // eslint-disable-line no-underscore-dangle
      ? res.body
      : res.body.replace(/"/g, '');
```

Comment (rwalle61) [reply packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/RequestPromiseResponse.js:7]: I noticed using `_json` would also work, but you choose which is clearer

Comment (ludeknovy) [reply packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/RequestPromiseResponse.js]: I like `._json` better, thanks for suggestion.

File: packages/jest-openapi/src/openapi-validator/lib/classes/RequestPromiseResponse.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ class RequestPromiseResponse extends AbstractResponse {
   constructor(res) {
     super(res);
     this.status = res.statusCode;
-    this.body = res.body.replace(/"/g, '');
+    this.body = res.request.headers.accept === 'application/json' ? res.body : res.body.replace(/"/g, '');
```
Comment (rwalle61) [inline packages/jest-openapi/src/openapi-validator/lib/classes/RequestPromiseResponse.js:7]: Thanks for the update to `chai-openapi-response-validator`. We should probably keep `jest-openapi` consistent by copying the update to here

**Comment by rwalle61** [comment]
Cool, shall I merge?

**Comment by ludeknovy** [comment]
not yet, fixing unit-tests :)

**Comment by ludeknovy** [comment]
@rwalle61 I hope now we should be good to merge.

**Comment by rwalle61** [comment]
Good catch! 

Thanks for this contribution, I'll try to release today

**Pull request merged**